core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/EntityScopeProviderImpl.java

```diff
@@ -108,20 +207,56 @@ public Scope getScopeForInterface(final Integer nodeId, final String ipAddress)
-                            .map("interface", "if-description", (i) -> Optional.ofNullable(i.getSnmpInterface()).map(OnmsSnmpInterface::getIfDescr))
-                            .map("interface", "phy-addr", (i) -> Optional.ofNullable(i.getSnmpInterface()).map(OnmsSnmpInterface::getPhysAddr))
+                    mapIpIntefaceKeys(new ObjectScope<>(ipInterface))
+                            .map(INTERFACE, "if-alias", (i) -> Optional.ofNullable(i.getSnmpInterface()).map(OnmsSnmpInterface::getIfAlias))
+                            .map(INTERFACE, "if-description", (i) -> Optional.ofNullable(i.getSnmpInterface()).map(OnmsSnmpInterface::getIfDescr))
+                            .map(INTERFACE, "phy-addr", (i) -> Optional.ofNullable(i.getSnmpInterface()).map(OnmsSnmpInterface::getPhysAddr))
             );
         });
 
         return metaDataScope;
     }
 
+    private static ObjectScope<OnmsIpInterface> mapIpIntefaceKeys(ObjectScope<OnmsIpInterface> scope) {
+        return scope.map(INTERFACE, "hostname", (i) -> Optional.ofNullable(i.getIpHostName()))
+                .map(INTERFACE, "address", (i) -> Optional.ofNullable(i.getIpAddress()).map(InetAddressUtils::toIpAddrString))
+                .map(INTERFACE, "netmask", (i) -> Optional.ofNullable(i.getNetMask()).map(InetAddressUtils::toIpAddrString))
+                .map(INTERFACE, "if-index", (i) -> Optional.ofNullable(i.getIfIndex()).map(Object::toString));
+    }
+
+    @Override
+    public Scope getScopeForInterfaceByIfIndex(final Integer nodeId, final int ifIndex) {
+        if (nodeId == null) {
+            return EmptyScope.EMPTY;
+        }
+
+        final Scope metaDataScope = this.sessionUtils.withReadOnlyTransaction(() -> {
+            final OnmsSnmpInterface snmpInterface = this.snmpInterfaceDao.findByNodeIdAndIfIndex(nodeId, ifIndex);
+            if (snmpInterface == null) {
+                return EmptyScope.EMPTY;
+            }
+
+            ArrayList<Scope> scopes = new ArrayList<>();
+
+            // SNMP interface facts
+            scopes.add(new ObjectScope<>(snmpInterface)
+                    .map(INTERFACE, "if-alias", (i) -> Optional.ofNullable(snmpInterface.getIfAlias()))
+                    .map(INTERFACE, "if-description", (i) -> Optional.ofNullable(snmpInterface.getIfDescr()))
+                    .map(INTERFACE, "phy-addr", (i) -> Optional.ofNullable(snmpInterface.getPhysAddr())));
+
+            // IP interface facts w/ meta-data extracted from IP interface
+            Optional.ofNullable(snmpInterface.getPrimaryIpInterface())
+                    .ifPresent(ipInterface -> {
+                        scopes.add(transform(ipInterface.getMetaData()));
+                        scopes.add(mapIpIntefaceKeys(new ObjectScope<>(ipInterface)));
+                    });
+
+            return new FallbackScope(scopes.toArray(new Scope[0]));
```

Comment:
`FallbackScope` uses a list internally. You can add a constructor which accepts a list (and alter the existing constructor to forward to the new one). If you use an `ImmutableList.Builder` here, this renders down to a no-op without any allocation.

Reply:
done.